server/src/main/java/org/elasticsearch/index/translog/TranslogHeader.java

```diff
@@ -0,0 +1,187 @@
+        } catch (CorruptIndexException | IndexFormatTooOldException | IndexFormatTooNewException e) {
+            tryReportOldVersionError(path, channel);
+            throw new TranslogCorruptedException("Translog header corrupted. path:" + path, e);
+        }
+        if (version == VERSION_CHECKSUMS) {
+            throw new IllegalStateException("pre-2.0 translog found [" + path + "]");
+        }
+        // Read the translogUUID
+        final int uuidLen = in.readInt();
+        if (uuidLen > channel.size()) {
+            throw new TranslogCorruptedException("uuid length can't be larger than the translog");
+        }
+        final BytesRef uuid = new BytesRef(uuidLen);
+        uuid.length = uuidLen;
+        in.read(uuid.bytes, uuid.offset, uuid.length);
+        final BytesRef expectedUUID = new BytesRef(translogUUID);
+        if (uuid.bytesEquals(expectedUUID) == false) {
+            throw new TranslogCorruptedException("expected shard UUID " + expectedUUID + " but got: " + uuid +
+                " this translog file belongs to a different translog. path:" + path);
+        }
+        // Read the primary term
+        final long primaryTerm;
+        if (version == VERSION_PRIMARY_TERM) {
+            primaryTerm = in.readLong();
+            assert primaryTerm >= 0 : "Primary term must be non-negative [" + primaryTerm + "]; translog path [" + path + "]";
+        } else {
+            assert version == VERSION_CHECKPOINTS : "Unknown header version [" + version + "]";
+            primaryTerm = UNKNOWN_PRIMARY_TERM;
+        }
+        // Verify the checksum
+        if (version >= VERSION_PRIMARY_TERM) {
+            Translog.verifyChecksum(in);
+        }
+        final int headerSizeInBytes = headerSize(version, uuid.length);
+        assert channel.position() == headerSizeInBytes :
+            "Header is not fully read; header size [" + headerSizeInBytes + "], position [" + channel.position() + "]";
+        return new TranslogHeader(translogUUID, primaryTerm, headerSizeInBytes);
+    }
+
+    private static void tryReportOldVersionError(final Path path, final FileChannel channel) throws IOException {
+        // Lucene's CodecUtil writes a magic number of 0x3FD76C17 with the header, in binary this looks like:
+        // binary: 0011 1111 1101 0111 0110 1100 0001 0111
+        // hex   :    3    f    d    7    6    c    1    7
+        //
+        // With version 0 of the translog, the first byte is the Operation.Type, which will always be between 0-4,
+        // so we know if we grab the first byte, it can be:
+        // 0x3f => Lucene's magic number, so we can assume it's version 1 or later
+        // 0x00 => version 0 of the translog
+        final byte b1 = Channels.readFromFileChannel(channel, 0, 1)[0];
+        if (b1 == 0x3f) { // LUCENE_CODEC_HEADER_BYTE
+            throw new TranslogCorruptedException("translog looks like version 1 or later, but has corrupted header. path:" + path);
```

Comment:
can you clarify what you mean with version 1 or later?

Reply:
It means that `the translog was created in ES 1.4 or later but corrupted`. Moved this from https://github.com/elastic/elasticsearch/blob/4ef3de40bca75f3b745c8d7adf60d8307a479bf2/server/src/main/java/org/elasticsearch/index/translog/TranslogReader.java#L110

Reply:
K. Thanks